File: Assignment_4/cpp/proxy/SongProxyServer.cpp
#include "SongProxyServer.h"
#include "SongServer.h"
#include "Song.h"

SongProxyServer::SongProxyServer(std::shared_ptr<SongServer> song_server)
{
  songServer = song_server;
}

Song * SongProxyServer::SearchById(int song_id)
{
  if (songServer)
  {
    songServer = std::make_shared<SongServer>();
  }

  auto song = songServer->SearchById(song_id);
  metadataCache.emplace_back(song->GetMetadata());

  return song;
}

std::vector<Song*> SongProxyServer::SearchByTitle(std::string title)
{
  if (songServer)
  {
    songServer = std::make_shared<SongServer>();
  }

  auto songs = songServer->SearchByTitle(title);

  for (auto s : songs)
  {
    metadataCache.emplace_back(s->GetMetadata());
  }

  return songs;
}

std::vector<Song*> SongProxyServer::SearchByAlbum(std::string album)
{
  if (songServer)
  {
    songServer = std::make_shared<SongServer>();
  }

  auto songs = songServer->SearchByAlbum(album);

  for (auto s : songs)
  {
    metadataCache.emplace_back(s->GetMetadata());
  }

  return songs;
}

void SongProxyServer::SetSongServer(std::shared_ptr<SongServer> song_server)
{
  songServer = song_server;
}

std::shared_ptr<SongServer> SongProxyServer::GetSongServer()
{
  return songServer;
}

std::vector<Metadata> SongProxyServer::GetMetadataCache()
{
  return metadataCache;
}

File: Assignment_4/java/proxy/src/SongProxyServer.java
import java.util.List;
import java.util.Vector;

public class SongProxyServer implements SongService {
    private Vector<Metadata> metadataCache;
    private SongService songServer = null;

    SongProxyServer() {
        metadataCache = new Vector<>();
    }

    @Override
    public Song searchById(Integer songID) {
        if (songServer == null) {
            songServer = new SongServer();
        }

        Song song = songServer.searchById(songID);
        metadataCache.add(song.GetMetadata());

        return song;
    }

    @Override
    public Vector<Song> searchByTitle(String title) {
        if (songServer == null) {
            songServer = new SongServer();
        }

        Vector<Song> song_list = songServer.searchByTitle(title);
        for (Song s : song_list) {
            if (s.GetMetadata().title == title) {
                metadataCache.add(s.GetMetadata());
            }
        }

        return song_list;
    }

    @Override
    public Vector<Song> searchByAlbum(String album) {
        if (songServer == null) {
            songServer = new SongServer();
        }

        Vector<Song> song_list = songServer.searchByAlbum(album);
        for (Song s : song_list) {
            if (s.GetMetadata().album == album) {
                metadataCache.add(s.GetMetadata());
            }
        }

        return song_list;
    }

    void PrintMetadata () {
        for (Metadata m : metadataCache) {
            System.out.println("Song ID: " + m.id);
            System.out.println("Song Title: " + m.title);
            System.out.println("Song Album: " + m.album);
            System.out.println("Song Duration: " + m.duration);
        }
    }
}
